[qz1004] iP

qz1004 opened the pull request:
# Jojo Bot
> "She's doing what she can." - Rosie

Jojo also goes by Jo. 
She frees your mind of having to remember things you need to do. She's
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2324S1/ip/archive/refs/heads/master.zip)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😁 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        new Jo(filePath).run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

qz1004 committed:
rename, greet, exit

qz1004 committed:
improved skeletal version so that it echos commands and exits when user types bye

qz1004 committed:
added the ability to store whatever text entered by the user and display them back to the user when requested

qz1004 committed:
added the ability to mark tasks as done and change the status back to not done

qz1004 committed:
added support for tracking three types of tasks: ToDos, Events, Deadlines

qz1004 committed:
test using the I/O redirection technique

qz1004 committed:
added exceptions to handle errors

qz1004 committed:
added support for deleting tasks from the list

qz1004 committed:
added enums

qz1004 committed:
Add ability to save tasks to hard drive

qz1004 committed:
Merge branch 'branch-Level-7'

qz1004 committed:
Add 'check' command to print tasks occurring on a specific date

qz1004 committed:
Merge branch 'branch-Level-8'

qz1004 committed:
Make code more OOP

qz1004 committed:
Merge branch 'A-MoreOOP'

qz1004 committed:
Divide classes into packages

qz1004 committed:
Merge branch 'add-gradle-support' of https://github.com/qz1004/ip into add-gradle-support

qz1004 committed:
Merge branch 'add-gradle-support'

qz1004 committed:
Add Gradle support

qz1004 committed:
Add JUnit tests for Parser and Task

qz1004 committed:
Add JAR file

qz1004 committed:
Add JavaDocs

qz1004 committed:
Tweak code to comply with coding standard

qz1004 committed:
Add JavaDocs

qz1004 committed:
Find a task by searching for a keyword

qz1004 committed:
Merge branch 'branch-Level-9'

qz1004 committed:
Merge branch 'branch-A-CodingStandard'

qz1004 committed:
Resolve merge conflicts

qz1004 committed:
Update code to comply with coding standards

junhonglow commented on src/main/java/jo/task/Task.java:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @return A string in the format: "T | Status | Description".
+     */
+    public String toFile() {
```
Perhaps it might be clearer if it was renamed as "toFileFormat" to explicitly indicate that is converting the task into a string representation for that purpose.

junhonglow commented on src/main/java/jo/TaskList.java:
```diff
@@ -0,0 +1,122 @@
+     * @param deadline The deadline to search for.
+     * @return An ArrayList of tasks matching the specified deadline and are not marked as done.
+     */
+    public ArrayList<Task> search(LocalDate deadline) {
```
I noticed that there is another method find() which sounds similar to search(). 
Perhaps a more descriptive method name would help to differentiate the two.

junhonglow reviewed (commented):
Overall, naming conventions were generally good with the exceptions of a few nitpicks.

qz1004 committed:
Create new file if file does not exist

qz1004 committed:
Add JavaFX GUI

qz1004 committed:
Merge branch 'branch-Level-10'

qz1004 committed:
Add ability to mark, unmark and delete multiple tasks using varargs

qz1004 committed:
Merge branch 'branch-A-Varargs'

Angelyxx commented on src/main/java/jo/Jo.java:
```diff
@@ -0,0 +1,85 @@
+ */
+public class Jo {
+
+    protected static String DEFAULT_FILE_PATH = "data/jo.txt";
```
Good use of capitalised snake_case for the constant DEFAULT_FILE_PATH

Angelyxx commented on src/main/java/jo/Storage.java:
```diff
@@ -0,0 +1,93 @@
+     * @throws JoException If an error occurs while reading the file or parsing tasks.
+     */
+    public ArrayList<Task> load() throws JoException {
+        ArrayList<Task> taskList = new ArrayList<>();
```
Use plural form 'tasks' when representing a collection of Task objects

Angelyxx commented on src/main/java/jo/Storage.java:
```diff
@@ -0,0 +1,93 @@
+                    taskList.add(new Deadline(task[2].trim(), isDone, LocalDate.parse(task[3].trim())));
+                } else if (taskType.equals("E")) {
+                    taskList.add(
+                            new Event(
+                                    task[2].trim(),
+                                    isDone,
+                                    LocalDate.parse(task[3].trim()),
+                                    LocalDate.parse(task[4].trim())
+                            )
```
Method name should be attached to the open parenthesis that follows it.

Angelyxx reviewed (commented):
LGTM, keep it up!

qz1004 committed:
Add error handling

Assertions are used to define assumptions about the program state so that the runtime can verify them. Exceptions are used to check if a user's input fall within a set of expected inputs and allows the program to react accordingly if it does not.

The Storage class assumes that all the Tasks stored in the data file has the indentifier "T", "D" or "E". Adding an assertion to check if this is true ensures that the program runs as desired.

The Parser class parse given inputs and call the corresponding methods to execute the specified commands. Adding an exception to check if commands that act on integers are indeed followed by digits.

qz1004 committed:
Improve code quality

qz1004 committed:
Merge pull request #1 from qz1004/branch-A-CodeQuality

Improve code quality

qz1004 committed:
Merge branch 'master' of https://github.com/qz1004/ip

qz1004 committed:
Merge branch 'master' into branch-A-Assertions

qz1004 committed:
Merge pull request #2 from qz1004/branch-A-Assertions

Add error handling

qz1004 committed:
Merge branch 'master' of https://github.com/qz1004/ip

qz1004 committed:
Fix mass operations feature

Due to errors in the code which went undiscovered in past commits, the feature to mass mark, unmark and delete raised runtime exceptions everytime it was called.

The issues lie in the parsing of the command input and not throwing an exception for unexpected inputs, which are fixed in this commit.

qz1004 committed:
Improve GUI

qz1004 committed:
Merge branch 'branch-A-BetterGui'

qz1004 committed:
Add user guide and finetune features

qz1004 committed:
Add user guide and finetune features

qz1004 committed:
Update user guide

qz1004 committed:
Add Ui.png

qz1004 committed:
Fix checkstyle

qz1004 committed:
Update gradle dependencies
